File: frontend/src/types/assessment.ts
import type { Question } from './question';

export type AssessmentStatus = 'draft' | 'published' | 'active' | 'closed' | 'archived';
export type InvitationStatus = 'pending' | 'sent' | 'opened' | 'started' | 'completed' | 'expired' | 'cancelled';
export type SessionStatus = 'not_started' | 'in_progress' | 'completed' | 'terminated' | 'timed_out';

export interface Assessment {
  id: string;
  organization_id: string;
  created_by_id: string;
  title: string;
  description: string | null;
  instructions: string | null;
  status: AssessmentStatus;
  time_limit_minutes: number | null;
  passing_score_pct: number;
  max_attempts: number;
  is_randomized: boolean;
  is_adaptive: boolean;
  show_score_immediately: boolean;
  proctoring_config: Record<string, unknown> | null;
  scheduled_start: string | null;
  scheduled_end: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface AssessmentSection {
  id: string;
  title: string;
  description: string | null;
  order_index: number;
  time_limit_minutes: number | null;
}

export interface AssessmentQuestion {
  id: string;
  question_id: string;
  section_id: string | null;
  order_index: number;
  weight: number;
  is_required: boolean;
  question: Question;
}

export interface AssessmentDetail extends Assessment {
  sections: AssessmentSection[];
  assessment_questions: AssessmentQuestion[];
}

export interface AssessmentListResponse {
  assessments: Assessment[];
  total: number;
  page: number;
  page_size: number;
}

export interface CandidateInvitation {
  id: string;
  assessment_id: string;
  organization_id: string;
  invited_by_id: string;
  candidate_email: string;
  candidate_name: string;
  token: string;
  status: InvitationStatus;
  sent_at: string | null;
  opened_at: string | null;
  expires_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface InvitationListResponse {
  invitations: CandidateInvitation[];
  total: number;
  page: number;
  page_size: number;
}

export interface CandidateResponseSummary {
  id: string;
  question_id: string;
  is_submitted: boolean;
  time_spent_seconds: number;
  final_score: number | null;
  max_score: number | null;
}

export interface CandidateSession {
  id: string;
  assessment_id: string;
  invitation_id: string;
  organization_id: string;
  candidate_email: string;
  candidate_name: string;
  status: SessionStatus;
  started_at: string | null;
  completed_at: string | null;
  time_remaining_seconds: number | null;
  current_question_index: number;
  total_score: number | null;
  total_max_score: number | null;
  score_pct: number | null;
  is_passed: boolean | null;
  proctoring_violations: number;
  ip_address: string | null;
  created_at: string;
  updated_at: string;
}

export interface SessionDetail extends CandidateSession {
  responses: CandidateResponseSummary[];
  proctoring_log: unknown[] | null;
}

export interface SessionListResponse {
  sessions: CandidateSession[];
  total: number;
  page: number;
  page_size: number;
}

File: backend/app/schemas/assessment.py
from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.question import QuestionResponse


class AssessmentSectionCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: Optional[str] = None
    order_index: int = 0
    time_limit_minutes: Optional[int] = None


class AssessmentSectionUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    order_index: Optional[int] = None
    time_limit_minutes: Optional[int] = None


class AssessmentSectionResponse(BaseModel):
    id: UUID
    title: str
    description: Optional[str]
    order_index: int
    time_limit_minutes: Optional[int]

    model_config = {"from_attributes": True}


class AssessmentQuestionAdd(BaseModel):
    question_id: UUID
    section_id: Optional[UUID] = None
    order_index: int = 0
    weight: float = 1.0
    is_required: bool = True


class AssessmentQuestionUpdate(BaseModel):
    section_id: Optional[UUID] = None
    order_index: Optional[int] = None
    weight: Optional[float] = None
    is_required: Optional[bool] = None


class AssessmentQuestionResponse(BaseModel):
    id: UUID
    question_id: UUID
    section_id: Optional[UUID]
    order_index: int
    weight: float
    is_required: bool
    question: QuestionResponse

    model_config = {"from_attributes": True}


class ReorderItem(BaseModel):
    id: UUID
    order_index: int


class AssessmentCreate(BaseModel):
    title: str = Field(min_length=1, max_length=500)
    description: Optional[str] = None
    instructions: Optional[str] = None
    time_limit_minutes: Optional[int] = None
    passing_score_pct: float = 50.0
    max_attempts: int = 1
    is_randomized: bool = False
    is_adaptive: bool = False
    show_score_immediately: bool = True
    proctoring_config: Optional[dict] = None
    scheduled_start: Optional[datetime] = None
    scheduled_end: Optional[datetime] = None


class AssessmentUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=500)
    description: Optional[str] = None
    instructions: Optional[str] = None
    time_limit_minutes: Optional[int] = None
    passing_score_pct: Optional[float] = None
    max_attempts: Optional[int] = None
    is_randomized: Optional[bool] = None
    is_adaptive: Optional[bool] = None
    show_score_immediately: Optional[bool] = None
    proctoring_config: Optional[dict] = None
    scheduled_start: Optional[datetime] = None
    scheduled_end: Optional[datetime] = None


class AssessmentResponse(BaseModel):
    id: UUID
    organization_id: UUID
    created_by_id: UUID
    title: str
    description: Optional[str]
    instructions: Optional[str]
    status: str
    time_limit_minutes: Optional[int]
    passing_score_pct: float
    max_attempts: int
    is_randomized: bool
    is_adaptive: bool
    show_score_immediately: bool
    proctoring_config: Optional[dict]
    scheduled_start: Optional[datetime]
    scheduled_end: Optional[datetime]
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class AssessmentDetailResponse(AssessmentResponse):
    sections: list[AssessmentSectionResponse] = []
    assessment_questions: list[AssessmentQuestionResponse] = []


class AssessmentListResponse(BaseModel):
    assessments: list[AssessmentResponse]
    total: int
    page: int
    page_size: int
